### PR TITLE
Clean all to review

### PR DESCRIPTION
**Pull request by manland**

### CODE REVIEW

File: server/activate.go [hunk 1]
```diff
@@ -1,191 +0,0 @@
...
-		}
-	}
-
-	p.cronSavePoison <- true
-
-	return nil
-}
-
-func (p *Plugin) ServeHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Request) {
-	switch r.URL.Path {
```
Comment (hanzei) [inline server/activate.go:68]: Might be worth to use https://github.com/gorilla/mux here.

File: server/analytic.go [hunk 2]
```diff
@@ -1,64 +0,0 @@
...
-
-type Analytic struct {
-	lock          sync.RWMutex
-	Start         time.Time
-	End           time.Time
-	Channels      map[string]int64
-	ChannelsReply map[string]int64
-	Users         map[string]int64
-	UsersReply    map[string]int64
-	FilesNb       int64
```
Comment (hanzei) [inline server/analytic.go:16]: Should this be `FilesMb`?

File: server/analytic.go [hunk 3]
```diff
@@ -1,64 +0,0 @@
-package main
-
-import (
-	"sync"
-	"time"
-)
-
-type Analytic struct {
```
Comment (hanzei) [inline server/analytic.go:8]: There should be a GoDoc comment for this struct and it fields.

File: server/analytic.go [hunk 4]
```diff
@@ -1,64 +0,0 @@
...
-		Channels:      make(map[string]int64),
-		ChannelsReply: make(map[string]int64),
-		Users:         make(map[string]int64),
-		UsersReply:    make(map[string]int64),
-		FilesNb:       int64(0),
-		FilesSize:     int64(0),
-	}
-}
-
-func (a *Analytic) Init() {
```
Comment (hanzei) [inline server/analytic.go:33]: Why is this needed if you have `NewAnalytic`?

File: server/analytic.go (hunk 3 above)
Comment (hanzei) [inline server/analytic.go:8]: If you have `NewAnalytic` this struct should be unexported.

File: server/message.go [hunk 5]
```diff
@@ -1,28 +0,0 @@
...
-	p.currentAnalytic.Channels[post.ChannelId]++
-	if post.ParentId != "" {
-		p.currentAnalytic.UsersReply[post.UserId]++
-		p.currentAnalytic.ChannelsReply[post.ChannelId]++
-	}
-}
-
-func (p *Plugin) FileWillBeUploaded(c *plugin.Context, info *model.FileInfo, file io.Reader, output io.Writer) (*model.FileInfo, string) {
-	p.currentAnalytic.WLock()
-	defer p.currentAnalytic.WUnlock()
```
Comment (hanzei) [inline server/message.go:24]: Nit: Add an empty line after this.

File: server/plugin.go [hunk 6]
```diff
@@ -1,304 +0,0 @@
...
-
-func (p *Plugin) ExecuteCommand(c *plugin.Context, args *model.CommandArgs) (*model.CommandResponse, *model.AppError) {
-	if !strings.HasPrefix(args.Command, "/"+CommandTrigger) {
-		return &model.CommandResponse{
-			ResponseType: model.COMMAND_RESPONSE_TYPE_EPHEMERAL,
-			Text:         fmt.Sprintf("Unknown command: " + args.Command),
-		}, nil
-	}
-
-	if strings.Contains(args.Command, "new") {
```
Comment (hanzei) [inline server/plugin.go:40]: Maybe split after the first space and use a `switch`  here?

File: server/plugin.go [hunk 7]
```diff
@@ -1,304 +0,0 @@
...
-		p.newSession()
-		allSessions, _ := p.allSessions()
-		j2, _ := json.Marshal(allSessions)
-		return &model.CommandResponse{
-			ResponseType: model.COMMAND_RESPONSE_TYPE_IN_CHANNEL,
-			Text:         string(j2),
```
Comment (hanzei) [inline server/plugin.go:46]: I'm not sure I get what you are doing here. Why would you post the whole data as a response?

File: server/plugin.go [hunk 8]
```diff
@@ -1,304 +0,0 @@
...
-	}
-
-	if strings.Contains(args.Command, "raw") {
-		j2, _ := json.Marshal(p.currentAnalytic)
-		return &model.CommandResponse{
-			ResponseType: model.COMMAND_RESPONSE_TYPE_IN_CHANNEL,
-			Text:         string(j2),
-		}, nil
-	}
-
```
Comment (hanzei) [inline server/plugin.go:66]: You should add a `help` command.

File: server/plugin.go [hunk 9]
```diff
@@ -1,304 +0,0 @@
...
-	data, err := p.prepareData()
```
Comment (hanzei) [inline server/plugin.go:67]: Should this be the default case? Why not use a separate command?

File: server/plugin.go [hunk 10]
```diff
@@ -1,304 +0,0 @@
...
-			upsert.nb = original.nb
-		}
-		if upsert.reply == 0 {
-			upsert.reply = original.reply
-		}
-	}
-	return append(originals, upsert)
-}
-
-func (p *Plugin) getChannelName(key string) (string, string, *model.AppError) {
```
Comment (hanzei) [inline server/plugin.go:254]: 1/5 that you should return `error` instead of `*model.AppError` here.

File: server/plugin.go [hunk 11]
```diff
@@ -1,304 +0,0 @@
...
-			DetailedError: err.Error(),
-		}
-	}
-	if channel.IsGroupOrDirect() {
-		return "", "Private", nil
-	}
-	return channel.Name, channel.DisplayName, nil
-}
-
-func (p *Plugin) getChannelDisplayName(key string) (string, *model.AppError) {
```
Comment (hanzei) [inline server/plugin.go:268]: :point_up:

File: server/plugin.go [hunk 12]
```diff
@@ -1,304 +0,0 @@
...
-		}
-	}
-	if channel.IsGroupOrDirect() {
-		return "Private", nil
-	}
-	return channel.DisplayName, nil
-}
-
-func (p *Plugin) getUsername(key string) (string, *model.AppError) {
-	user, err := p.API.GetUser(key)
```
Comment (hanzei) [inline server/plugin.go:283]: :point_up:

File: server/plugin.go [hunk 13]
```diff
@@ -1,304 +0,0 @@
...
-	if err != nil {
-		return "", &model.AppError{
-			Message:       "Can't retreive user name",
-			DetailedError: err.Error(),
-		}
-	}
-	return user.Username, nil
-}
-
-func byteCountDecimal(b int64) string {
```
Comment (hanzei) [inline server/plugin.go:293]: Maybe add unit test?

File: server/save.go [hunk 14]
```diff
@@ -1,83 +0,0 @@
...
-				p.save()
-				return
-			}
-		}
-	}()
-}
-
-func (p *Plugin) save() {
-	p.currentAnalytic.RLock()
-	defer p.currentAnalytic.RUnlock()
```
Comment (hanzei) [inline server/save.go:39]: Nit: Add a linebreak

File: server/save.go [hunk 15]
```diff
@@ -1,83 +0,0 @@
...
-	if err != nil {
-		p.API.LogWarn("can't get all sessions", "err", err.Error())
-	}
-
-	j2, err2 := json.Marshal(append(allAnalytics, p.currentAnalytic.Close()))
-	if err2 != nil {
-		p.API.LogWarn("can't marshal internal analytics data")
-	}
-	if err := p.API.KVSet("allAnalytics", j2); err != nil {
-		p.API.LogError("failed to unmarshal analytics from kv use new one", "err", err.Error())
```
Comment (hanzei) [inline server/save.go:80]: Copy/Past error :wink

File: server/save.go [hunk 16]
```diff
@@ -1,83 +0,0 @@
...
-				p.save()
-			case <-poison:
-				p.save()
-				return
-			}
-		}
-	}()
-}
-
-func (p *Plugin) save() {
```
Comment (hanzei) [inline server/save.go:37]: Maybe rename this to `saveCurrentAnalytic`?

File: server/save.go [hunk 17]
```diff
@@ -1,83 +0,0 @@
...
-	allAnalytics, err := p.allSessions()
-	if err != nil {
-		p.API.LogWarn("can't get all sessions", "err", err.Error())
-	}
-
-	j2, err2 := json.Marshal(append(allAnalytics, p.currentAnalytic.Close()))
-	if err2 != nil {
-		p.API.LogWarn("can't marshal internal analytics data")
-	}
-	if err := p.API.KVSet("allAnalytics", j2); err != nil {
```
Comment (hanzei) [inline server/save.go:79]: I don't understand why you use two key. What is the difference between `allAnalytics` and `analytics`?

**Review by hanzei — COMMENTED** [review]
One meta comment: I think you should create a `util` package and move all you util methods in it.

**Comment by hanzei** [comment]
I leaved a bunch of comments. Some of them are out of scope, but you could create issues/PR's for them.

Overall I really like this plugin. Keep up the good work! :+1:

File: server/plugin.go (hunk 6 above)
Comment (manland) [reply server/plugin.go:40]: Sorry it was for debug purpose, I have deleted all this commands in https://github.com/manland/mattermost-plugin-analytics/pull/3/files#diff-7e42e9629fece6c3b56011e7de76423bL40

File: server/plugin.go (hunk 7 above)
Comment (manland) [reply server/plugin.go:46]: Sorry it was for debug purpose, I have deleted all this commands in https://github.com/manland/mattermost-plugin-analytics/pull/3/files#diff-7e42e9629fece6c3b56011e7de76423bL40

File: server/plugin.go (hunk 8 above)
Comment (manland) [reply server/plugin.go:66]: Now /analytics do only one thing : return analytics. I think an help command is not necessary  https://github.com/manland/mattermost-plugin-analytics/pull/3/files#diff-7e42e9629fece6c3b56011e7de76423bL40

File: server/plugin.go (hunk 9 above)
Comment (manland) [reply server/plugin.go:67]: Again, sorry it was for debug purpose, I have deleted all this commands in https://github.com/manland/mattermost-plugin-analytics/pull/3/files#diff-7e42e9629fece6c3b56011e7de76423bL40

File: server/plugin.go (hunk 10 above)
Comment (manland) [reply server/plugin.go:254]: :heavy_check_mark:

File: server/plugin.go (hunk 11 above)
Comment (manland) [reply server/plugin.go:268]: :heavy_check_mark:

File: server/plugin.go (hunk 12 above)
Comment (manland) [reply server/plugin.go:283]: :heavy_check_mark:

File: server/save.go (hunk 14 above)
Comment (manland) [reply server/save.go:39]: :heavy_check_mark:

File: server/save.go (hunk 15 above)
Comment (manland) [reply server/save.go:80]: :heavy_check_mark:

File: server/save.go (hunk 16 above)
Comment (manland) [reply server/save.go:37]: :heavy_check_mark:

File: server/save.go (hunk 17 above)
Comment (manland) [reply server/save.go:79]: `analytics` is the current session of analytics, `allAnalytics` is an array with all sessions. `analytics` is loaded when plugin start, stored every minute in case of failure. The second is loaded when a report is requested to display all data : 

selection_186

File: server/message.go (hunk 5 above)
Comment (manland) [reply server/message.go:24]: :heavy_check_mark:

File: server/analytic.go (hunk 2 above)
Comment (manland) [reply server/analytic.go:16]: No it's the number of files downloaded, the next property of this struct `FilesSize` represent the weight of all files.

File: server/analytic.go (hunk 4 above)
Comment (manland) [reply server/analytic.go:33]: Good catch :D It is a hack because when I create a new session I hold a reference to the currentAnalytic. I need to invistage after all PR are merged https://github.com/manland/mattermost-plugin-analytics/issues/17

File: server/analytic.go (hunk 2 above)
Comment (hanzei) [reply server/analytic.go:16]: For what does `Nb` stand for?

File: server/analytic.go (hunk 3 above)
Comment (manland) [reply server/analytic.go:8]: I'm not sure
selection_194

Comment (manland) [reply server/analytic.go:8]: :heavy_check_mark:

File: server/analytic.go (hunk 2 above)
Comment (manland) [reply server/analytic.go:16]: As the future godoc say ( :D ) : `FilesNb store number of files uploaded`

selection_197

selection_198

I'm sorry my english is not fluent :sweat_smile: maybe better name can be found

**Comment by manland** [comment]
I think i have resolved all this issues in https://github.com/manland/mattermost-plugin-analytics/pull/18

> One meta comment: I think you should create a `util` package and move all you util methods in it.

Yes definitively, I need to make packages. Maybe a `chart` package and `report`

> Overall I really like this plugin. Keep up the good work! +1

Thank you very much for your time.

I start next PR #3
